### ISSUE TITLE
Numerous issues with install process. SSH on server no longer accessible

### ISSUE THREAD

**Issue by inspiredearth** [carlalexander/debops-wordpress#171]
I've came across numerous issues in the installation process. I have no idea why, and there is very little in the WIki instructions to indicate what the issues are.

I found I had to use the following command for a start, otherwise numerous permissions errors were the result.
`sudo debops bootstrap -u jonathan -b --become-method=sudo -K --key-file=~/.ssh/id_rsa`

Anything else, wouldn't work. If you want more details, let me know.
The command seemed to complete successfully after that, with this final result:

`www.domain.nz  : ok=15   changed=8    unreachable=0    failed=0`

Then I ran the `debops` command:

It finished with these errors:

```
TASK [debops.pki : Check Ansible Controller bash version] **********************************************************************************************************************************************
fatal: [www.domain.nz -> localhost]: FAILED! => {"changed": false, "cmd": ["/usr/bin/env", "bash", "-c", "echo $BASH_VERSION"], "delta": "0:00:00.016270", "end": "2017-07-09 22:15:00.757165", "failed": true, "rc": -6, "start": "2017-07-09 22:15:00.740895", "stderr": "dyld: Library not loaded: /usr/local/opt/readline/lib/libreadline.6.dylib\n  Referenced from: /usr/local/bin/bash\n  Reason: image not found", "stderr_lines": ["dyld: Library not loaded: /usr/local/opt/readline/lib/libreadline.6.dylib", "  Referenced from: /usr/local/bin/bash", "  Reason: image not found"], "stdout": "", "stdout_lines": []}

NO MORE HOSTS LEFT *************************************************************************************************************************************************************************************
 [WARNING]: Could not create retry file '/Users/jonathan/Library/Application Support/debops/debops-playbooks/playbooks/site.retry'.         [Errno 13] Permission denied:
u'/Users/jonathan/Library/Application Support/debops/debops-playbooks/playbooks/site.retry'
```
```
PLAY RECAP *********************************************************************************************************************************************************************************************
www.domain.nz  : ok=51   changed=20   unreachable=0    failed=1
```

I then proceeded with the `debops wordpress` command.
First attempt, it had this error:

```
TASK [Gathering Facts] *********************************************************************************************************************************************************************************
fatal: [www.domain.nz]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: Permission denied (publickey,password).\r\n", "unreachable": true}
	to retry, use: --limit @/Users/jonathan/development/git/debops-wordpress/playbooks/wordpress.retry

PLAY RECAP *********************************************************************************************************************************************************************************************
www.domain.nz  : ok=0    changed=0    unreachable=1    failed=0
```
I checked the SSH connection to the server and it was fine. So I tried the command again. This time it worked, but resulted in these errors:

```
TASK [debops.users : DebOps pre_tasks hook] ************************************************************************************************************************************************************
fatal: [www.domain.nz]: FAILED! => {"failed": true, "msg": "lookup plugin (task_src) not found"}

RUNNING HANDLER [debops.ferm : Restart ferm] ***********************************************************************************************************************************************************
 [WARNING]: Could not create retry file '/Users/jonathan/development/git/debops-wordpress/playbooks/wordpress.retry'.         [Errno 13] Permission denied: u'/Users/jonathan/development/git/debops-
wordpress/playbooks/wordpress.retry'


PLAY RECAP *********************************************************************************************************************************************************************************************
www.domain.nz  : ok=44   changed=25   unreachable=0    failed=1
```

I see that this last error is in fact mentioned in the Troubleshooting page on the Wiki. So I was going to follow the advice there. But I am now unable to access the server with SSH. Something in the above processes has made SSH access impossible.

I get this error every time:
`ssh_exchange_identification: read: Connection reset by peer`

When I go to the Reset Password (also for resetting the SSH key) on Azure, I see this warning: `The VM agent is either unavailable, or not installed, which may prevent VMAccess from running.
`

So it would appear the VM Agent was wiped out by the debops installation process.
So, I thought I'd fire up the Bash terminal Azure provides in the Dashboard. But it gives me this error:

```
You have no storage mounted
Azure Cloud Shell requires an Azure file share to persist files. Learn more 
This will create a new storage account for you and this will incur a small monthly cost. View pricing
```
This is after restarting the server.
So it would appear the debops installation process also screwed the boot process of the VM. Looks like I'll have to kill it, and start from scratch.

I would like to know what you suggest as a way forward.

Thanks,

Jonathan

**Comment by drybjed** [comment]
Hi. First of all, you ran a command which returned with an error, and you proceeded to the next one? There shouldn't be any errors in the deployment process, if there are, something is wrong. You should first fix the current error before running the next deployment commands.

Knowing more about your environment (both the remote host and the Ansible Controller) whould certainly help with diagnosing the issue. From your post I know that you are using Azure for deployment, but what do you use as the Ansible Controller? Is it MacOS X?

Now, about the bootstrap command:
```
sudo debops bootstrap -u jonathan -b --become-method=sudo -K --key-file=~/.ssh/id_rsa
```
You said about "numerous permission errors" - permissions to what, to read files on the Ansible Controller, to access the remote host over SSH?

The command is executed using `sudo`, so you are switching from your regular user account to the `root` account, on your Ansible Controller. Then, you tell Ansible to use your own SSH key (I assume that shell expanded the `~/.ssh/id_rsa` path to the absolute path before handling the arguments to `sudo`) because the `root` user's `~/.ssh/id_rsa` key, where SSH looks for by default, either doesn't exist, or is not the one deployed on the remote host. If you drop the `sudo` command and run the `debops` command on your own user account, it should correctly access your own SSH key.

The rest of the `debops bootstrap` command is fine - everyone's environment is different and will be accessed differently. The `bootstrap` playbook exists so that it should help you unify the host access, so you will be able to login to your own account over SSH and `sudo` to `root` without a password.

Now, the `debops` command... And now I see that you are using a MacOS X. And because you downloaded the roles using `sudo debops-update` instead of `debops-update`, all the playbook directories are using wrong owner/group, ie. not your own. And because Ansible creates its `.retry` files at the same location where the playbook is by default, and it can't in your case, it exits with an error.

And Ansible exited with an error because the task from the `debops.pki` role that checks if the `bash` version on Ansible Controller is supported, exited with:
```
dyld: Library not loaded: /usr/local/opt/readline/lib/libreadline.6.dylib\n  Referenced from: /usr/local/bin/bash\n  Reason: image not found
```
The `dyld` is some kind of dynamic linker on MacOS X. No idea how to fix that, I don't use this platform. Anyway, you should first fix your environment, especially `/usr/local/bin/bash` execution, before going further with `debops wordpress` command. Good luck. :-)

**Comment by inspiredearth** [comment]
@drybjed Thanks Maciej, I appreciate the input.
So, my ignorance around the consequences of some of my actions made a bit of a mess of things.
I trashed the old VM, and have a fresh one hot off the press. This time I'll get it right.

Yes, the deploying machine (Ansible Controller) is MacOS (10.12.5). If there's anything more I can tell you about the local environment, please let me know.

The remote environment is on Azure, and is a fresh installation of Ubuntu 16.04 LT Server.

Is it okay for me to post the errors that come up from precisely following the instructions, so that we can resolve them one by one? Hopefully, if I iron out the permissions issues that came up for the `debops bootstrap -u jonathan` I may in fact avoid any of the later errors. Let's see.
Local

UPDATE: Here is what I discovered, which I think was the cause of the errors I was getting when I run the first command, the first time around. It turned out there was an issue with bash, which I gather was installed by Homebrew. Homebrew has not be upgrading it when doing a `brew update` command, so it was an older version. I found others online reporting the same issue. It was causing this error:
```
dyld: Library not loaded: /usr/local/opt/readline/lib/libreadline.6.dylib
  Referenced from: /usr/local/bin/bash
  Reason: image not found
Abort trap: 6
```
From memory, that (along with some related permissions errors) was the error I was getting when running `debops bootstrap -u jonathan`. What I now realise is that by running it as sudo, it was likely using the bash installed with MacOS, not the one Homebrew has installed into my user account.

**Comment by inspiredearth** [comment]
So far I've gone through step one and two without errors. I notice quite a few things are skipped. Is that to be expected, or should I consider that a problem? For example,
```
TASK [debops.auth : Configure pam-mkhomedir for LDAP authentication] *********************************************
skipping: [www.domain.nz]

TASK [debops.auth : Install packages require for LDAP authentication] ********************************************
skipping: [www.domain.nz] => (item=[])

TASK [debops.auth : Divert original /etc/nslcd.conf] *************************************************************
skipping: [www.domain.nz]

TASK [debops.auth : Check if host password hash exists] **********************************************************
skipping: [www.domain.nz]

TASK [debops.auth : Read hash of host password] ******************************************************************
skipping: [www.domain.nz]

TASK [debops.auth : Generate host password] **********************************************************************
skipping: [www.domain.nz]

TASK [debops.auth : Save hash of host password] ******************************************************************
skipping: [www.domain.nz]

TASK [debops.auth : Make sure that host base DN exists] **********************************************************
skipping: [www.domain.nz]

TASK [debops.auth : Make sure that host account exists] **********************************************************
skipping: [www.domain.nz]

TASK [debops.auth : Set host bind password in LDAP] **************************************************************
skipping: [www.domain.nz] => (item=(censored due to no_log))

TASK [debops.auth : Configure nslcd for LDAP authentication] *****************************************************
skipping: [www.domain.nz]

TASK [debops.auth : Restart nslcd if configuration changed] ******************************************************
skipping: [www.domain.nz]
```

**Comment by drybjed** [comment]
I'm glad that you made it work. :-) Skipping some tasks is fine, some of the DebOps roles like `debops.auth` have support for more advanced functions like LDAP in your example. If you don't enable LDAP support explicitly, they will be skipped.

**Comment by inspiredearth** [comment]
One other things I am not clear about in your comment, and I wish to make I am not overlooking something to do with it...
> And because you downloaded the roles using sudo debops-update instead of debops-update, all the playbook directories are using wrong owner/group, ie. not your own.

At what point does `debops-update` come into it? Or is this something that one of the three commands does as part of its routine?

**Comment by drybjed** [comment]
If you followed @carlalexander's install guide to the letter, it tells you to download the DebOps roles using the command:

    sudo debops-update

I'm not sure how MacOS X environment resolves this, but apparently the roles and playbooks were downloaded on your user account (I think) but with `root` as the file/directory owner.

Then, when you ran the `debops` command which failed, Ansible tried to create a `.retry` file for the failed playbook, in the same directory that the playbook resides in. But, since that directory was owned by `root` and not your own user, Ansible couldn't create that file and existed with the permission error.

@carlalexander, you might consider changing the command in the install guide to `debops-update`, to avoid such issues in the future. DebOps is meant to be run from an unprivileged user account on the Ansible Controller.

NB: DebOps roles and playbooks can be installed system-wide for easy sharing in multi-user environment. However, the issue with Ansible creating the `.retry` files in the location of the playbook instead of the user's `$HOME` directory may cause problems. This can be disabled in the Ansible configuration file, and perhaps DebOps should do that by default.

**Comment by inspiredearth** [comment]
Ah okay. Back when I was setting up the OSX as the Ansibal controller. (https://github.com/carlalexander/debops-wordpress/wiki/Installation#install-ansible-and-debops)

Got it. Thank fully I've had no errors so far (the `debops` command is still running), so no retry files have been created (or attempted).

**Comment by inspiredearth** [comment]
In the end I got one error with the `debops wordpress` command.
```
TASK [debops.users : DebOps pre_tasks hook] *******************************************************************************************************************************************************************************************************
fatal: [www.domain.nz]: FAILED! => {"failed": true, "msg": "lookup plugin (task_src) not found"}
 [WARNING]: Could not create retry file '/Users/jonathan/development/git/debops-wordpress/playbooks/wordpress.retry'.         [Errno 13] Permission denied: u'/Users/jonathan/development/git/debops-wordpress/playbooks/wordpress.retry'


PLAY RECAP ****************************************************************************************************************************************************************************************************************************************
www.domain.nz  : ok=44   changed=17   unreachable=0    failed=1
```
Any suggestions on what to do about this?

**Comment by drybjed** [comment]
Some of the DebOps roles use a custom Ansible lookup plugins which rely on the `debops` Python module for their configuration. If you installed Ansible using Homebrew and DebOps using PyPI, Ansible from Homebrew does not have access to the PyPI modules. I think that removing the Homebrew version and installing the PyPI version should fix this issue.

**Comment by drybjed** [comment]
Oh, you also need to point Ansible to the lookup plugins which are contained in the `debops-playbooks` repository. Make sure that `ansible.cfg` config file has that directory in the `lookup_plugins` path variable.

**Comment by inspiredearth** [comment]
Not sure I fully understand.
Python was installed with homebrew, as per the instructions. `brew install python`.

Ansible and DebOps were installed via python, within the command `sudo pip install ansible debops passlib pycrypto`

So I am not sure what PyPI is referring to.

**Comment by inspiredearth** [comment]
ansible.cfg has the following setting for lookup_plugins:

`lookup_plugins = "/Users/jonathan/development/git/debops-wordpress/ansible/lookup_plugins":"/Users/jonathan/Library/Application Support/debops/debops-playbooks/playbooks/lookup_plugins":"/usr/share/ansible/lookup_plugins"`

I notice only one of those paths exists on my system. `/Users/jonathan/
Library/Application Support/debops/debops-playbooks/playbooks/lookup_plugins`

I see there is actually no `/ansible/` folder in my debops-wordpress folder. In fact there isn't one on my system (apart from all the ansible folders in `/Users/jonathan//Library/Application Support/debops/debops-playbooks/roles/debops....` template paths.

So I am wondering if Ansible even installed? Could I have got this far without it?

**Comment by inspiredearth** [comment]
So, I tried doing an install of Ansible again. I recall now, that I had permissions errors last time. Same thing this time:

```
The directory '/Users/jonathan/Library/Caches/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
The directory '/Users/jonathan/Library/Caches/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
Requirement already satisfied: ansible in /usr/local/lib/python2.7/site-packages
Requirement already satisfied: jinja2 in /usr/local/lib/python2.7/site-packages (from ansible)
Requirement already satisfied: PyYAML in /usr/local/lib/python2.7/site-packages (from ansible)
Requirement already satisfied: paramiko in /usr/local/lib/python2.7/site-packages (from ansible)
Requirement already satisfied: pycrypto>=2.6 in /usr/local/lib/python2.7/site-packages (from ansible)
Requirement already satisfied: setuptools in /usr/local/lib/python2.7/site-packages (from ansible)
Requirement already satisfied: MarkupSafe in /usr/local/lib/python2.7/site-packages (from jinja2->ansible)
Requirement already satisfied: bcrypt>=3.1.3 in /usr/local/lib/python2.7/site-packages (from paramiko->ansible)
Requirement already satisfied: cryptography>=1.1 in /usr/local/lib/python2.7/site-packages (from paramiko->ansible)
Requirement already satisfied: pynacl>=1.0.1 in /usr/local/lib/python2.7/site-packages (from paramiko->ansible)
Requirement already satisfied: pyasn1>=0.1.7 in /usr/local/lib/python2.7/site-packages (from paramiko->ansible)
Requirement already satisfied: six>=1.4.1 in /usr/local/lib/python2.7/site-packages (from bcrypt>=3.1.3->paramiko->ansible)
Requirement already satisfied: cffi>=1.1 in /usr/local/lib/python2.7/site-packages (from bcrypt>=3.1.3->paramiko->ansible)
Requirement already satisfied: idna>=2.0 in /usr/local/lib/python2.7/site-packages (from cryptography>=1.1->paramiko->ansible)
Requirement already satisfied: enum34 in /usr/local/lib/python2.7/site-packages (from cryptography>=1.1->paramiko->ansible)
Requirement already satisfied: ipaddress in /usr/local/lib/python2.7/site-packages (from cryptography>=1.1->paramiko->ansible)
Requirement already satisfied: pycparser in /usr/local/lib/python2.7/site-packages (from cffi>=1.1->bcrypt>=3.1.3->paramiko->ansible)
```

I recall I used the -H option last time, so I did that again now.
However, there's still no ansible folder at all but one of the paths mentioned above.

`pip list` shows that `ansible (2.3.1.0)` and `debops (0.4.4)` and `passlib (1.6.5)` and `pycrypto (2.6.1)` are all installed, as per the instructions.

**Comment by drybjed** [comment]
How about this. Try running in the debops-wordpress project directory (where the `ansible.cfg` file is);

    debops service/users

This executest the `debops.users` role which uses the `task_src` lookup plugin. Does it pass correctly?

**Comment by inspiredearth** [comment]
Trying that now.

In the mean time I took ownership of all files/folders in debops-wordpress folder. As I am not aware of any reason for them to be owned by root.

RESULT:
Didn't pass. 
I see it's now trying to create the in a completely different location now, which again has generated a permission error. I am not sure if that matters, in terms of whether it has any relevance to the success of the command. I suspect not.

```
TASK [debops.users : DebOps pre_tasks hook] *******************************************************************************************
fatal: [www.domain.nz]: FAILED! => {"failed": true, "msg": "lookup plugin (task_src) not found"}
 [WARNING]: Could not create retry file '/Users/jonathan/Library/Application Support/debops/debops-playbooks/playbooks/service/users.retry'.         [Errno 13] Permission denied: u'/Users/jonathan/Library/Application Support/debops/debops-
playbooks/playbooks/service/users.retry'


PLAY RECAP *******************************************************************************************
www.domain.nz  : ok=1    changed=0    unreachable=0    failed=1
```

**Comment by inspiredearth** [comment]
@drybjed I am still trying to resolve this lookup_plugin issue. Coming back to the path that's set for those... Does Ansible need to find those plugins in **all** the locations that are specified, or just any one of them? 

Again, the three paths specified are:
`/Users/jonathan/development/git/debops-wordpress/ansible/lookup_plugins":"/Users/jonathan/Library/Application Support/debops/debops-playbooks/playbooks/lookup_plugins":"/usr/share/ansible/lookup_plugins`

The lookup_plugins do exist at `/Users/jonathan/Library/Application Support/debops/debops-playbooks/playbooks/lookup_plugins`.

I created an `ansible` folder in `debops-wordpress`, and copied the `lookup_plugins` into there too. 

I edited the lookup_plugins setting in the `.debops.cfg` file to remove the reference to /usr/share/ansible/lookup_plugins, since macOS doesn't allow content to be added to the `/usr/` folder, so it shouldn't even be in the picture for such an operation.

So the files are now located at both paths Ansible is meant to look for them, and yet the Ansible script is throwing out an error that they are not there.

Is it possible it's looking for specific plugins in that folder and they are not there, or is it just complaining that the folder itself doesn't exist?

Any suggestions?

**Comment by drybjed** [comment]
They can be in just 1 location, can be in many locations. The Ansible `*_plugins` variables work like a `$PATH` environment variable, first found plugin wins.

I'm not sure why it doesn't work for you... Does use of the `debops` playbook by itself produce any errors? Did you try `debops service/users` and has that worked as well?

**Comment by inspiredearth** [comment]
Yeah, I am not sure either.
As mentioned, `debops service/users` is throwing the same error.

```
j-mbp:debops-wordpress jonathan$ debops service/users
Running Ansible playbook from:
/Users/jonathan/Library/Application Support/debops/debops-playbooks/playbooks/service/users.yml ...

PLAY [Manage local users and groups] *******************************************************************************************

TASK [Gathering Facts] *******************************************************************************************
ok: [www.domain.nz]

TASK [debops.secret : Create secret directories on Ansible Controller] *******************************************************************************************

TASK [debops.users : DebOps pre_tasks hook] *******************************************************************************************
fatal: [www.domain.nz]: FAILED! => {"failed": true, "msg": "lookup plugin (task_src) not found"}
 [WARNING]: Could not create retry file '/Users/jonathan/Library/Application Support/debops/debops-playbooks/playbooks/service/users.retry'.         [Errno 13] Permission denied: u'/Users/jonathan/Library/Application Support/debops/debops-
playbooks/playbooks/service/users.retry'

PLAY RECAP *******************************************************************************************
www.domain.nz  : ok=1    changed=0    unreachable=0    failed=1
```

It's starting to look like I just have to abort this 4 day mission. Although that would be such a pity, after spending so much time on it.

**Comment by drybjed** [comment]
You could try setting up a Debian VM (Jessie is most likely to work correctly as the Ansible Controller at the moment) and run DebOps from there. This should work fine, since it's the project's native environment.

**Comment by inspiredearth** [comment]
So, create a Debian VM and set it up as the Ansible controller, and then deploy from there to the Ubuntu VM ?

**Comment by drybjed** [comment]
Yes, exactly. You could even use DebOps to configure it if you wish, make sure that you have access to the console in case of trouble. Root password will be stored in the `secret/` directory.

**Comment by inspiredearth** [comment]
To use DebOps to configure it, I gather I can just work my way through the instructions at https://github.com/debops/debops-tools ?

I see you're the originator of DebOps. Thanks for assisting. I guess that means if you don't know why my debops-wordpress install isn't working, then there isn't much chance anyone else will. If Yoda tells a Jedi in training that his light-saber is broken beyond repair, then best he accept that, and create a new one.  :-)

So, with that  in mind, I have deployed a Debian VM. And I will use DebOps to configure it.

**Comment by drybjed** [comment]
:-)

You can check out the [DebOps Getting Started Guide](https://docs.debops.org/en/latest/debops-playbooks/docs/guides/getting-started.html), this version is newer and includes better examples, althoug it's sometimes best to consul the roles as well, project is moving at a fast pace.

**Comment by inspiredearth** [comment]
I've decided to skip using DebOps to configure the Debian VM. It's proving too much to get my head around, after days of trying unsuccessfully to get DebOps / Ansible working on my Mac. So I am simply following the standard instructions for getting debops-wordpress set up on the VM that will now (hopefully) be the Ansible controller.

During the execution of this command, `sudo pip install ansible debops passlib pycrypto`

I got this error:
```
Cleaning up...
Command python setup.py egg_info failed with error code 1 in /tmp/pip-build-HSLw0k/pynacl
Storing debug log for failure in /root/.pip/pip.log
```

When I break the `sudo pip install ansible debops passlib pycrypto` down to installing them one at a time, it is the ansible installation that causes the error. DebOps installs fine, as do passlib and pycrypto.

Is this something to be concerned about, or can I proceed with the debops-wordpress instuctions?

Would it be a good idea to remove the pip install of ansible (and debops) and install them with apt-get?

**Comment by drybjed** [comment]
Try installing Ansible from GitHub checkout, not from PyPI. There were issues with this release and DebOps in the past, I just stick to source.

[This script](https://github.com/debops/ansible-debops/blob/master/files/script/bootstrap-ansible.sh) should work fine. Download it, make it executable and run:

    ./bootstrap-ansible.sh stable-2.3

This should install all of the necessary dependencies from Debian, download Ansible source and build a `.deb` package from the latest `stable-2.3` branch and install it. After that you should be good to go.

The current Ansible version in Jessie and Stretch is too old for DebOps, hopefully a backported one will be available at some point. And with this script, you can easily replace the installed version with a different one if needed.

**Comment by inspiredearth** [comment]
I successfully installed Ansible 1.7.2 using Deb package manager. Would it be best to more that, and install via the script from Git?

**Comment by drybjed** [comment]
Yes, the Ansible v1.x series is not supported for a long time now.

**Comment by inspiredearth** [comment]
I was able to upgrade it to 2.3.1 via this repository... `deb http://ppa.launchpad.net/ansible/ansible/ubuntu trusty main`
Is that okay, or is it necessary to use the sh file and build a deb locally from the files downloaded from Github?

**Comment by drybjed** [comment]
Well, this `.deb` package was compiled for Ubuntu Trusty and not for Debian. It's your choice if you want to trust it or not.

**Comment by inspiredearth** [comment]
Ok. Sounds like I should go with the script. Doing that now. Should I uninstall the version that I installed and updated on Deb apt-get, or is that not required?

Also, do I run the script as su ?

**Comment by drybjed** [comment]
The built package should replace the currently installed one, you don't need to specifically uninstall it.

**Comment by inspiredearth** [comment]
Okay. Well, here's hoping this all works.
` ansible --version` shows

> ansible 2.3.2.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.9 (default, Jun 29 2016, 13:08:31) [GCC 4.9.2]

`pip show debops` gives me:

> Name: debops
Version: 0.4.4
Location: /usr/local/lib/python2.7/dist-packages
Requires: netaddr, argparse, passlib

and `pip list` presents:

> **ansible (2.3.2.0)**
argparse (1.2.1)
cffi (0.8.6)
chardet (2.3.0)
colorama (0.3.2)
cryptography (0.6.1)
**debops (0.4.4)**
docutils (0.12)
ecdsa (0.11)
html5lib (0.999)
httplib2 (0.9)
Jinja2 (2.7.3)
MarkupSafe (0.23)
ndg-httpsclient (0.3.2)
netaddr (0.7.19)
nose (1.3.4)
paramiko (1.15.1)
passlib (1.7.1)
pip (1.5.6)
ply (3.4)
pyasn1 (0.1.7)
pycparser (2.10)
pycrypto (2.6.1)
Pygments (2.0.1)
pyOpenSSL (0.14)
PyYAML (3.11)
requests (2.4.3)
roman (2.0.0)
setuptools (5.5.1)
six (1.8.0)
Sphinx (1.2.3)
urllib3 (1.9.1)
wheel (0.24.0)
wsgiref (0.1.2)

Is that all looking about right?

**Comment by drybjed** [comment]
Yes. :-) Now on your user account, run `debops-update` which should get all the roles and playbooks and put them in ``/.local/share/debops/`. After that you can set up a new DebOps project directory with:

    debops-init ~/src/projects/project1

**Comment by inspiredearth** [comment]
Fine to do all that without sudo ?
Just double checking, because I know the original instructions (https://github.com/carlalexander/debops-wordpress/wiki/Installation) had me doing everything with sudo.

**Comment by inspiredearth** [comment]
Also, is the debops-init necessary for the debops-wordpress project, or just jump into:

>  git clone https://github.com/carlalexander/debops-wordpress.git
$ cd debops-wordpress/
$ git checkout tags/v0.4.0

**Comment by drybjed** [comment]
Yes, you should do all `debops` commands on your own user account, so without `sudo`.

**Comment by drybjed** [comment]
You could get into `debops-wordpress`, sure - it's basically the same as a result of `debops-init`.

**Comment by inspiredearth** [comment]
Well... It looks like after four days of messing around with this, I might be almost there. Just one more question, if you don't mind. I notice that `debopts-init` creates a project folder with `/ansible` folder in it. And that the debops-wordpress project folder doesn't have an `/ansible` folder in it. Is that to be expected?

**Comment by drybjed** [comment]
It depends on the configuration provided by `ansible.cfg` config file. The `debops` script generates that file, and it should detect if the `inventory/` directory is in an `ansible/` subdirectory or not. Having it in a subdirectory helps organize your work directory, so I prefer it as a subdirectory.

**Comment by inspiredearth** [comment]
Okay. Thanks. Makes sense.
I am just working on a slight hitch. Now that I am using a remote VM as the Ansible Controller, I have to somehow give that machine SSH access to the Azure VM. But, it would seem Azure VMs only support storing one public key file for SSH access. 
There's always a catch! :-) 

SOLVED.

**Comment by inspiredearth** [comment]
It took my many hours to get SSH access working from the Debian workstation to the Ubuntu Server. So I am now resuming the process of deploying debops-wordpress. I have a few questions.

Previously I did the `debops bootstrap -u` with the username `jonathan`. But I gather I won't be able to do that from the Debian machine, as I had to set up a different user on there, with a different keypair. When I connect from Deb w/s to Ubuntu srv it's with a different user and key than when I connect from the Mac computer. Obviously, that means each ends up in a different home directory on the Ubuntu server.

Does any of this have any impact on deploying DebOps?

I notice the Ubuntu server now says " This system is managed by Ansible" when I SSH into it. So I gather the failed attempts at deploying ansible/DebOps from the Mac did in fact achieve something on that server. Does that need to be undone before I attempt things again from the Debian w/s ?

Basically, I can't afford any more lost time with issues in this deployment (It's absorbed the best part of 5 days now), so I 'd like to ensure it goes right from here in.

**Comment by drybjed** [comment]
> But I gather I won't be able to do that from the Debian machine, as I had to set up a different user on there, with a different keypair

Why? It's your own VM, correct? You should be able to add your own account without issues. Using a different keypair might be a good idea, sure.

Remember that you can have multiple SSH keys in the `~/.ssh/authorized_keys` file on the remote host, so you could add the VM key to your existing account, or create a new account and add both of your SSH keys there.

DebOps controls user access via system groups, `admins`, `sshusers`. Since you created a new admin account, make sure that it's present in the `admins` group. Otherwise you might not be able to access the host via SSH or switch accoounts via `sudo`.

If you haven't copied the `secret/` directory (DebOps project directory as a whole would be best though), DebOps will generate new internal CA and certificates on the remote host will be re-signed with it. If the VM's incoming IP address is different than your Mac IP address, it might get added to the list of Ansible Controllers managed by the `debops.core` role, and firewall/TCP wrappers should get updated. Otherwise all of the software installed and configuration set on the remote host that DebOps is aware of should just be checked for consistency.

Make sure that you have console access to the Ubuntu VM. If you set the password on your user account, it shouldn't be changed by DebOps.

**Comment by inspiredearth** [comment]
> Why? It's your own VM, correct? You should be able to add your own account without issues. Using a different keypair might be a good idea, sure.

The Debian machine has the user I access it with, and then the username/key pair it uses to access the Ubuntu server. It's the latter that I figured might have some impact on things. Should I have just added the public key from that machine to the original user I had set up on the Ubuntu server? So that it's logging in over SSH with the same Ubuntu username?
The reason I ended up doing it the way I did is because Azure documentation, forum comments, Azure Portal Interface, etc., all gave me the impression Azure VMs were (strangely) limited to one user/key pair for SSH access. Long story.

> DebOps controls user access via system groups, admins, sshusers. Since you created a new admin account, make sure that it's present in the admins group. Otherwise you might not be able to access the host via SSH or switch accoounts via sudo.

Yeah, that's pretty much why the above because a long story. I had a hell of a time figuring out why the new user couldn't access the Ubuntu Srv with SSH. Eh.

> Make sure that you have console access to the Ubuntu VM. If you set the password on your user account, it shouldn't be changed by DebOps.

I am a relatively noob to all this. Please clarify what this last point means. By console access are you referring to SSH access or Azure Portal access? I assume "console" is the Az Portal?

By "set the password on your user account" are you referring to setting a password through Azure Portal? Or just accessing the terminal on the server via SSH and running the command there to set a password for my account?

What I gather from your reply, is that it should be fine to go ahead and run `debops bootstrap -u MYSSHUSER` from the Debian VM.

Although, to keep things streamlined, before doing that I am going to add the pub key from the keypair on the Deb VM to the original user's `authorized_keys` file on the Ubuntu VM, so that I can SSH from Deb to Ubuntu with the same user that I have been SSHing into it from Mac.

**Comment by drybjed** [comment]
Yes, "console access" usually means physical access to the host where you can switch to a TTY and login directly, it can also mean out of band access to a VM over VNC connection, which is usually available in VPS control panels. Basically non-SSH access to the host's terminal.

You can just set the password using the `passwd` command after logging in to your user account. Check afterwards that you can login using console. You can even stay logged in during the deployment to make sure everything works.

**Comment by inspiredearth** [comment]
Whilst on the Ubuntu srv (to which Ansible has been deployed (at least in part, from the attempts made via my Mac), I notice there's something odd with the bash. For example, on both user accounts (the original one, and the one I set up for accessing it from Deb [which I'll soon delete]) I can't use the `ls` command, unless I add an option to it. Just `ls` on its own does nothing. I have to do `ls -a` or `ls -l` to get any output.

I noticed it earlier on when I was logged in with the second username, and I just assumed it was something broken in the configuration of that user. But I see now it occurs with the new user too. It would appear this is something the Ansible deployment has done. Have you got any ideas why it might be? (before I bring it up on stack exchange to seek help with it).

**Comment by drybjed** [comment]
Umm. Your home directory does not have any "normal" files, just some hidden ones with a dot at the beggining of the filename. Since there are no normal files to show, `ls` doesn't output anything as any UNIX-like command is tend to do. `ls -l` tells you that there are 0 files, correct? And `ls -a` shows all files present in the directory, including the "hidden" ones with a dot at the beginning.

Try this: create a directory with `mkdir testdir`, or create an empty file with `touch testfile`. Then run `ls` and see what happens.

**Comment by inspiredearth** [comment]
That's bizarre. I could have sworn it was occurring no matter what directory I was in. But maybe each time I tested it I was in yet another directory with only hidden files. So yes, I now see it's only in the home directories, which have only hidden . files (and anywhere else with only . files, or none at all). Sorry to have bothered you with a ridiculous question. [cringing).
Back to the deployment.

**Comment by inspiredearth** [comment]
I think I am close to throwing in the towel on this. When I run `debops bootstrap -u jonathan` I get the following error:

```
$ debops bootstrap -u jonathan
Running Ansible playbook from:
/home/jonathan/.local/share/debops/debops-playbooks/playbooks/bootstrap.yml ...

PLAY [Bootstrap Python support on a host] 

TASK [debops.bootstrap/raw : Update APT repositories, install mandatory packages] *******************************************************************************************
fatal: [www.DOMAIN.nz]: FAILED! => {"changed": true, "failed": true, "rc": 100, "stderr": "Shared connection to www.DOMAIN.nz closed.\r\n", "stdout": "\rReading package lists... 0%\r\rReading package lists... 100%\r\rReading package lists... Done\r\r\nW: chmod 0700 of directory /var/lib/apt/lists/partial failed - SetupAPTPartialDirectory (1: Operation not permitted)\r\nE: Could not open lock file /var/lib/apt/lists/lock - open (13: Permission denied)\r\nE: Unable to lock directory /var/lib/apt/lists/\r\nW: Problem unlinking the file /var/cache/apt/pkgcache.bin - RemoveCaches (13: Permission denied)\r\nW: Problem unlinking the file /var/cache/apt/srcpkgcache.bin - RemoveCaches (13: Permission denied)\r\nE: Could not open lock file /var/lib/dpkg/lock - open (13: Permission denied)\r\nE: Unable to lock the administration directory (/var/lib/dpkg/), are you root?\r\n", "stdout_lines": ["", "Reading package lists... 0%", "", "Reading package lists... 100%", "", "Reading package lists... Done", "", "W: chmod 0700 of directory /var/lib/apt/lists/partial failed - SetupAPTPartialDirectory (1: Operation not permitted)", "E: Could not open lock file /var/lib/apt/lists/lock - open (13: Permission denied)", "E: Unable to lock directory /var/lib/apt/lists/", "W: Problem unlinking the file /var/cache/apt/pkgcache.bin - RemoveCaches (13: Permission denied)", "W: Problem unlinking the file /var/cache/apt/srcpkgcache.bin - RemoveCaches (13: Permission denied)", "E: Could not open lock file /var/lib/dpkg/lock - open (13: Permission denied)", "E: Unable to lock the administration directory (/var/lib/dpkg/), are you root?"]}
	to retry, use: --limit @/home/jonathan/.local/share/debops/debops-playbooks/playbooks/bootstrap.retry

PLAY RECAP 
www.DOMAIN.nz  : ok=0    changed=0    unreachable=0    failed=1
```
Should I just pack it in, and cut my losses, or is this something obvious I can fix?

**Comment by drybjed** [comment]
You connected to the user account but didn't tell Ansible to use `sudo`. The `bootstrap` playbook doesn't do this automatically. Append `--become` or `-b` to the command line arguments.

**Comment by drybjed** [comment]
Oh, and if `sudo` requires password, add `-K`.

**Comment by inspiredearth** [comment]
So I need to do what I figured out back at the start of this mission...
`debops bootstrap -u jonathan -b --become-method=sudo -K --key-file=~/.ssh/id_rsa` (minus the `sudo` I originally at the front of all that)  ?
It seems to me the instructions at https://github.com/carlalexander/debops-wordpress/wiki/Configuring-your-server need updating or elaborating.

**Comment by inspiredearth** [comment]
Okay, with that command, plus the password I set for `jonathan` a little earlier on, the bootstrap is ticking along. Fingers crossed. Thanks so much for all your help. I really do appreciate it. I am setting all this up for a non-profit, to help them out. It's been a huge shit-arse of a mission, and I would have given up days ago had you not so willing assisted with all my pleads for information and help.

**Comment by drybjed** [comment]
`--become-method=sudo` is redundant since Ansible automatically selects `sudo` if it's detected on the host (IIRC). The `--key-file=~/.ssh/id_rsa` should be redundant because this is the default private key file SSH command uses, at least on my system. So you should be fine with running it like:

    debops bootstrap -u jonathan -b -K

**Comment by inspiredearth** [comment]
So far so good...
`PLAY RECAP **********************************************************************************************
www.DOMAIN.nz  : ok=14   changed=2    unreachable=0    failed=0`

**Comment by drybjed** [comment]
Great, now check if you can login and `sudo su` without password.

Also, this seems to be a documentation problem, I guess a better first steps guide with explanation of command line arguments and what they are used for is needed, since there are so many different environments, both as Ansible Controllers and remote hosts.

**Comment by drybjed** [comment]
On the bright side, when everything finally goes correctly you will be able to use DebOps for other things as well. :-)

**Comment by inspiredearth** [comment]
That result was just from the `debops bootstrap` command. Now running `debops` (so far with lots of beautiful green in the results, and no red). 
yes, I can `sudu su` without entering a password (on the Ubuntu srv, user `jonathan`

**Comment by inspiredearth** [comment]
Lol....
Yep. I freak'n hope so! lol
I've learnt and remembered (as it was about 12 years ago I was last working with linux, other than dabbling in the command line on OS X from time to time) a lot in the process. I keep telling myself, "Hang in there... this will all come in real handy..."

**Comment by inspiredearth** [comment]
I've never before been so enamoured—thrilled is probably another descritpion—at the site of green, turquoise, and yellow text scrolling up on a screen. And so pissed when I see a line of red!! :-)

**Comment by inspiredearth** [comment]
`debops` command successful.
`www.DOMAIN.nz  : ok=212  changed=12   unreachable=0    failed=0`

Next... `debops wordpress`
Got one error:

```
RUNNING HANDLER [debops.fail2ban : Restart fail2ban] *******************************************************************************************
fatal: [www.DOMAIN.nz]: FAILED! => {"changed": false, "failed": true, "msg": "Unable to restart service fail2ban: Job for fail2ban.service failed because the control process exited with error code. See \"systemctl status fail2ban.service\" and \"journalctl -xe\" for details.\n"}
```

Any thoughts Maestro?

Looks like fail2ban had a fail2restart.

This (https://github.com/carlalexander/debops-wordpress/issues/123) might be a related issue.

**Comment by inspiredearth** [comment]
Ran a second time, with the `debops wordpress --limit @/home/jonathan/debops-wordpress/playbooks/wordpress.retry` and it was successful.

Does this mean I might of actually made it to the end of this process?

**Comment by inspiredearth** [comment]
Looks like fail2ban is not operational on the Server:
```
$sudo service fail2ban status
● fail2ban.service - Fail2Ban Service
   Loaded: loaded (/lib/systemd/system/fail2ban.service; enabled; vendor preset: enabled)
   Active: inactive (dead) (Result: exit-code) since Wed 2017-07-12 10:30:26 UTC; 16min ago
     Docs: man:fail2ban(1)
 Main PID: 73079 (code=killed, signal=TERM)

Jul 12 10:30:26 www systemd[1]: fail2ban.service: Control process exited, code=exited status=255
Jul 12 10:30:26 www systemd[1]: Failed to start Fail2Ban Service.
Jul 12 10:30:26 www systemd[1]: fail2ban.service: Unit entered failed state.
Jul 12 10:30:26 www systemd[1]: fail2ban.service: Failed with result 'exit-code'.
Jul 12 10:30:26 www systemd[1]: fail2ban.service: Service hold-off time over, scheduling restart.
Jul 12 10:30:26 www systemd[1]: Stopped Fail2Ban Service.
Jul 12 10:30:26 www systemd[1]: fail2ban.service: Start request repeated too quickly.
Jul 12 10:30:26 www systemd[1]: Failed to start Fail2Ban Service.
```

I see a few services are installed but not running. No idea if that's to be expected: (I suspect it is)


```
$ sudo service --status-all
 [ + ]  acpid
 [ + ]  apparmor
 [ + ]  apport
 [ + ]  atd
 [ - ]  bootmisc.sh
 [ - ]  checkfs.sh
 [ - ]  checkroot-bootclean.sh
 [ - ]  checkroot.sh
 [ + ]  console-setup
 [ + ]  cron
 [ - ]  cryptdisks
 [ - ]  cryptdisks-early
 [ + ]  dbus
 [ - ]  fail2ban
 [ + ]  ferm
 [ + ]  grub-common
 [ + ]  haveged
 [ - ]  hostname.sh
 [ - ]  hwclock.sh
 [ + ]  irqbalance
 [ + ]  iscsid
 [ + ]  keyboard-setup
 [ - ]  killprocs
 [ + ]  kmod
 [ - ]  lvm2
 [ + ]  lvm2-lvmetad
 [ + ]  lvm2-lvmpolld
 [ + ]  lxcfs
 [ - ]  lxd
 [ + ]  mdadm
 [ - ]  mdadm-waitidle
 [ - ]  mountall-bootclean.sh
 [ - ]  mountall.sh
 [ - ]  mountdevsubfs.sh
 [ - ]  mountkernfs.sh
 [ - ]  mountnfs-bootclean.sh
 [ - ]  mountnfs.sh
 [ + ]  mysql
 [ + ]  networking
 [ + ]  nginx
 [ + ]  ondemand
 [ + ]  open-iscsi
 [ - ]  open-vm-tools
 [ + ]  php7.0-fpm
 [ - ]  plymouth
 [ - ]  plymouth-log
 [ + ]  postfix
 [ + ]  procps
 [ + ]  rc.local
 [ + ]  redis-server
 [ + ]  resolvconf
 [ - ]  rsync
 [ + ]  rsyslog
 [ - ]  screen-cleanup
 [ - ]  sendsigs
 [ + ]  ssh
 [ - ]  thermald
 [ + ]  udev
 [ + ]  ufw
 [ - ]  umountfs
 [ - ]  umountnfs.sh
 [ - ]  umountroot
 [ + ]  unattended-upgrades
 [ + ]  unscd
 [ + ]  urandom
 [ - ]  uuidd
 [ + ]  varnish
 [ + ]  varnishlog
 [ + ]  varnishncsa
```

**Comment by inspiredearth** [comment]
Perhaps this shows up something useful:

```$ journalctl -r -u fail2ban.service
-- Logs begin at Tue 2017-07-11 17:03:47 UTC, end at Wed 2017-07-12 10:57:04 UTC. --
Jul 12 10:56:15 www systemd[1]: Failed to start Fail2Ban Service.
Jul 12 10:56:15 www systemd[1]: fail2ban.service: Start request repeated too quickly.
Jul 12 10:56:15 www systemd[1]: Stopped Fail2Ban Service.
Jul 12 10:56:15 www systemd[1]: fail2ban.service: Service hold-off time over, scheduling restart.
Jul 12 10:56:15 www systemd[1]: fail2ban.service: Failed with result 'exit-code'.
Jul 12 10:56:15 www systemd[1]: fail2ban.service: Unit entered failed state.
Jul 12 10:56:15 www systemd[1]: Failed to start Fail2Ban Service.
Jul 12 10:56:15 www systemd[1]: fail2ban.service: Control process exited, code=exited status=255
Jul 12 10:56:15 www fail2ban-client[100884]: ERROR  Failed during configuration: Have not found any log file for wordpress-badbots jail
Jul 12 10:56:15 www fail2ban-client[100884]: ERROR  No file(s) found for glob /var/log/nginx/varnish*access.log
Jul 12 10:56:14 www systemd[1]: fail2ban.service: Service hold-off time over, scheduling restart.```

**Comment by inspiredearth** [comment]
A server reboot seems to have fixed the fail2ban issue.

Wordpress doesn't load though. So I am now investigating how to resolve that.

**Comment by drybjed** [comment]
I haven't touched `fail2ban` for a bit, I suppose that you could comment it out for now in the `debops-wordpress` playbook.

**Comment by inspiredearth** [comment]
Fail2ban ended up being okay after I restarted the server.
The issue now is that Wordpress doesn't load. Connection times out.

**Comment by drybjed** [comment]
Is `nginx` up? Is `php-fpm` up? Run:
```
netstat -nlp
ls -lR /run/php*
```
Anything interesting in the `nginx` logs? Look in `/var/log/nginx/`.

**Comment by inspiredearth** [comment]
Yep, both nginx and php7.0-fpm are up.

netstat -nlp:

```
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.0.1:3306          0.0.0.0:*               LISTEN      -
tcp        0      0 127.0.0.1:6379          0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:25              0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:6081            0.0.0.0:*               LISTEN      -
tcp        0      0 127.0.0.1:6082          0.0.0.0:*               LISTEN      -
tcp6       0      0 :::80                   :::*                    LISTEN      -
tcp6       0      0 :::22                   :::*                    LISTEN      -
tcp6       0      0 :::25                   :::*                    LISTEN      -
tcp6       0      0 :::6081                 :::*                    LISTEN      -
udp        0      0 0.0.0.0:68              0.0.0.0:*                           -
Active UNIX domain sockets (only servers)
Proto RefCnt Flags       Type       State         I-Node   PID/Program name    Path
unix  2      [ ACC ]     STREAM     LISTENING     550282   40177/systemd       /run/user/1000/systemd/private
unix  2      [ ACC ]     SEQPACKET  LISTENING     11020    -                   /run/udev/control
unix  2      [ ACC ]     STREAM     LISTENING     17863    -                   /var/run/nscd/socket
unix  2      [ ACC ]     STREAM     LISTENING     17864    -                   /var/run/.nscd_socket
unix  2      [ ACC ]     STREAM     LISTENING     24003    -                   /var/run/fail2ban/fail2ban.sock
unix  2      [ ACC ]     STREAM     LISTENING     23527    -                   /run/php7.0-fpm-wordpress.sock
unix  2      [ ACC ]     STREAM     LISTENING     10957    -                   /run/systemd/private
unix  2      [ ACC ]     STREAM     LISTENING     10961    -                   /run/lvm/lvmetad.socket
unix  2      [ ACC ]     STREAM     LISTENING     10963    -                   /run/systemd/journal/stdout
unix  2      [ ACC ]     STREAM     LISTENING     10972    -                   /run/systemd/fsck.progress
unix  2      [ ACC ]     STREAM     LISTENING     11013    -                   /run/lvm/lvmpolld.socket
unix  2      [ ACC ]     STREAM     LISTENING     31804    -                   public/pickup
unix  2      [ ACC ]     STREAM     LISTENING     31808    -                   public/cleanup
unix  2      [ ACC ]     STREAM     LISTENING     31811    -                   public/qmgr
unix  2      [ ACC ]     STREAM     LISTENING     31815    -                   private/tlsmgr
unix  2      [ ACC ]     STREAM     LISTENING     31892    -                   private/rewrite
unix  2      [ ACC ]     STREAM     LISTENING     31895    -                   private/bounce
unix  2      [ ACC ]     STREAM     LISTENING     31898    -                   private/defer
unix  2      [ ACC ]     STREAM     LISTENING     31901    -                   private/trace
unix  2      [ ACC ]     STREAM     LISTENING     31904    -                   private/verify
unix  2      [ ACC ]     STREAM     LISTENING     31907    -                   public/flush
unix  2      [ ACC ]     STREAM     LISTENING     31910    -                   private/proxymap
unix  2      [ ACC ]     STREAM     LISTENING     31913    -                   private/proxywrite
unix  2      [ ACC ]     STREAM     LISTENING     31916    -                   private/smtp
unix  2      [ ACC ]     STREAM     LISTENING     31919    -                   private/relay
unix  2      [ ACC ]     STREAM     LISTENING     31922    -                   public/showq
unix  2      [ ACC ]     STREAM     LISTENING     31925    -                   private/error
unix  2      [ ACC ]     STREAM     LISTENING     31928    -                   private/retry
unix  2      [ ACC ]     STREAM     LISTENING     31931    -                   private/discard
unix  2      [ ACC ]     STREAM     LISTENING     31943    -                   private/local
unix  2      [ ACC ]     STREAM     LISTENING     31946    -                   private/virtual
unix  2      [ ACC ]     STREAM     LISTENING     31949    -                   private/lmtp
unix  2      [ ACC ]     STREAM     LISTENING     31952    -                   private/anvil
unix  2      [ ACC ]     STREAM     LISTENING     31955    -                   private/scache
unix  2      [ ACC ]     STREAM     LISTENING     31958    -                   private/maildrop
unix  2      [ ACC ]     STREAM     LISTENING     31961    -                   private/uucp
unix  2      [ ACC ]     STREAM     LISTENING     31970    -                   private/ifmail
unix  2      [ ACC ]     STREAM     LISTENING     31973    -                   private/bsmtp
unix  2      [ ACC ]     STREAM     LISTENING     31976    -                   private/scalemail-backend
unix  2      [ ACC ]     STREAM     LISTENING     31979    -                   private/mailman
unix  2      [ ACC ]     STREAM     LISTENING     16077    -                   /run/snapd.socket
unix  2      [ ACC ]     STREAM     LISTENING     16811    -                   @ISCSIADM_ABSTRACT_NAMESPACE
unix  2      [ ACC ]     STREAM     LISTENING     16081    -                   /var/lib/lxd/unix.socket
unix  2      [ ACC ]     STREAM     LISTENING     16078    -                   /run/snapd-snap.socket
unix  2      [ ACC ]     STREAM     LISTENING     16079    -                   /run/uuidd/request
unix  2      [ ACC ]     STREAM     LISTENING     16080    -                   /run/acpid.socket
unix  2      [ ACC ]     STREAM     LISTENING     16082    -                   /var/run/dbus/system_bus_socket
unix  2      [ ACC ]     STREAM     LISTENING     43350    -                   /var/run/mysqld/mysqld.sock
```

Looks to me like Port 80 is only being listened to over TCP6. Shouldn't TCP also being listening on 80?

ls -lR /run/php*
```
srw-rw---- 1 www-data www-data  0 Jul 12 11:11 /run/php7.0-fpm-wordpress.sock

/run/php:
total 4
-rw-r--r-- 1 root root 4 Jul 12 11:11 php7.0-fpm.pid
```
(No idea what to make of that)

nginx errorlog shows:
`2017/07/12 11:13:22 [emerg] 3560#3560: ipv6only is not supported on addr "0.0.0.0:8080", ignored in /etc/nginx/sites-enabled/www.DOMAIN.nz.conf:10`

_Which I gather is related to the issue pointed out above, that only IPV6 is listening on port 80. Using Ansible to make the changes, how would I go about making TCP listen on port 80?_


Aside from the WP-no-go issue, I note that all configuration files say they are generated remotely, by Ansible. Can you point me in the right direction for learning how to make changes to configuration files, via Ansible?

UPDATE:  [[ I've since figure out I do everything in the `../inventory/hosts` file.

I suspect the downside to having deployed Debops-wordpress from a remote workstation is I don't have easy access to the configuration files. I'll look into sharing them and mounting locally. ]]

**Comment by drybjed** [comment]
No, you shouldn't do everything in the `ansible/inventory/hosts` file, but I'll get back to that in a minute.

`nginx` listens only on IPv6 because it does that by default on Debian/Ubuntu. The `debops.nginx` role should have configured it to listen on IPv4 as well. Run this:

    grep 'ipv6only' /etc/nginx/sites-enabled/*

It should return at least two lines, one for `80` port and one for `443`. If it does, restart `nginx`:

    systemctl restart nginx.service

If it doesn't, remove the `/etc/ansible/facts.d/nginx.fact` file, this action will tell the `debops.nginx` role to re-evaluate all configuration and select a default host. Re-run `debops` commands and check for `ipv6only` again. If it doesn't work, let me know.

As for configuration changes - DebOps roles are controlled by Ansible inventory, but you don't want to put the variables in the main `hosts` INI file, because they are too complex for that. Use YAML syntax and put files in `inventory/group_vars/` and/or `inventory/host_vars/` directories. You can lookup the variables for a given role in the documentation. Here's [an example DebOps inventory](https://github.com/debops/examples/tree/master/devlab), check the contents of the `hosts` file and look how various files and directories are laid out. The contents are pretty old at this point, so don't rely specifically on the variables being correct.

If you want to view all the variables at once, you can run `debops-defaults` command which will aggregate all of the `defaults/main.yml` files from DebOps roles and put them in a read-only file for you. Don't copy everything to the inventory though, just the variables you want to modify. You can also grep the output like this:

    debops-defaults | grep nginx

**Comment by inspiredearth** [comment]
Sorry, I see I forgot to mention... I'd taken a look in the nginx conf file, and it is set to only listen on port 8080, and in the varnish conf file it's listening to port 80.  Here's the confirmation:

```
/etc/nginx/sites-enabled/varnish.www.DOMAIN.nz.conf:        listen [::]:80 default_server ipv6only=off;
/etc/nginx/sites-enabled/www.DOMAIN.nz.conf:        listen 8080 default_server ipv6only=off;
```
So it would seem it somehow didn't get configured correctly.
I've removed the `nginx.fact file. I assume I run `debops` from the Ansible Controller, correct?

**Comment by drybjed** [comment]
Ah, right, it's `debops-wordpress`, it has `nginx` behind `varnish`. Then check if `varnish` is working correctly. I don't have experience with it, perhaps @carlalexander could be of a help here.

Yes, you run `debops` commands from the Ansible Controller.

**Comment by inspiredearth** [comment]
Yep,  doing above tricks didn't change the configuration.

I am beginning to see, however, just how powerful and handy DebOps and Ansible are. I've been looking through the configuration docs, and it's pretty cool.

Varnish is up. It's logs don't show any issues. Actually, it's log is empty.

**Comment by drybjed** [comment]
You could try shutting down the firewall and talking to `nginx` directly, see if it works:
```
systemctl stop ferm
curl http://<website>:8080/
```

**Comment by inspiredearth** [comment]
Thanks for the suggestion.

Unfortunately even with the firewall stopped, the curl to the domain (also to the IP) on port 8080 just hangs there, receiving nothing.

As I don't know the cause, I have no idea what the solution is. I am pretty close to just having to ditch this and go another route, as the sites this server is meant to be hosting are meant to be going online first thing next week. :-(

Should I post a new issue thread, for just this particular issue? What do you suggest? It seems I have successfully going through all the build and deployment steps, free from errors, and yet something is obviously not right.

**Comment by inspiredearth** [comment]
is there no way to manually add in the necessary config parameters so that port 80 is being listened via TCP, by whatever it is that should be listening to it? A suspect it's Varnish, right? And then if varnish receives a request for something it's not caching (or able to cache) it then parses the request through to Nginx. Or something like that, right?

**Comment by drybjed** [comment]
Sure, the Ansible runs aren't automated. After you configured a host with DebOps, you can mess in the config files as you see fit. Just be aware that if you run DebOps roles on it again, your changes will be overwritten - if you want to keep them, after you fixed whatever issue was causing it, add the necessary configuration via Ansible inventory variables so that subsequent DebOps runs will apply the changes correctly.

**Comment by inspiredearth** [comment]
I ran `wget http://www.domain.nz:8080` and got:

```
--2017-07-14 09:35:39--  http://www.domain.nz:8080/
Resolving www.domain.nz (www.domain.nz)... 127.0.1.1
Connecting to www.domain.nz (www.domain.nz)|127.0.1.1|:8080... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: http://www.domain.nz/ [following]
--2017-07-14 09:35:40--  http://www.domain.nz/
Connecting to www.domain.nz (www.domain.nz)|127.0.1.1|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 52122 (51K) [text/html]
index.html: Permission denied

Cannot write to ‘index.html’ (Success).
```

Should those `permission denied` results be a cause for concern?

I had `varnishlog` running whilst I did that. It didn't produce any "errors".

**Comment by drybjed** [comment]
`wget` couldn't create a new file in the current directory. Are you an owner of this directory?

**Comment by inspiredearth** [comment]
Ah. Okay. That's what that means. No, I happened to be doing it from a varnish related directory.

**Comment by drybjed** [comment]
Are you sure that firewall allows connections to ports `80` and `443`? Check using:

    iptables -v -L

**Comment by inspiredearth** [comment]
```
--2017-07-14 09:43:01--  http://www.domain.nz:8080/
Resolving www.domain.nz (www.domain.nz)... 127.0.1.1
Connecting to www.domain.nz (www.domain.nz)|127.0.1.1|:8080... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: http://www.domain.nz/ [following]
--2017-07-14 09:43:01--  http://www.domain.nz/
Connecting to www.domain.nz (www.domain.nz)|127.0.1.1|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 52122 (51K) [text/html]
Saving to: ‘index.html’

index.html                 100%[=======================================>]  50.90K  --.-KB/s    in 0s

2017-07-14 09:43:01 (173 MB/s) - ‘index.html’ saved [52122/52122]
```

**Comment by inspiredearth** [comment]
Currently I have shut down `fail2ban` and `ferm`, and I was assuming that would mean there was no firewall in effect.

**Comment by inspiredearth** [comment]
iptables...

```
Chain INPUT (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     all  --  lo     any     anywhere             anywhere
   41  2360 ansible_controller-rules  tcp  --  any    any     anywhere             anywhere             tcp dpt:ssh
    0     0 hashlimit-icmp-flood  icmp --  any    any     anywhere             anywhere
    0     0 DROP       all  --  any    any     anywhere             anywhere             ctstate INVALID
  105 49300 ACCEPT     all  --  any    any     anywhere             anywhere             ctstate RELATED,ESTABLISHED
    0     0 REJECT     all  --  any    any     anywhere             anywhere             recent: UPDATE seconds: 600 name: fail2ban-wordpress-login side: source mask: 255.255.255.255 reject-with icmp-admin-prohibited
    0     0 REJECT     all  --  any    any     anywhere             anywhere             recent: UPDATE seconds: 7200 name: fail2ban-wordpress-ddos side: source mask: 255.255.255.255 reject-with icmp-admin-prohibited
    0     0 REJECT     all  --  any    any     anywhere             anywhere             recent: UPDATE seconds: 7200 name: fail2ban-wordpress-badbots side: source mask: 255.255.255.255 reject-with icmp-admin-prohibited
    0     0 REJECT     all  --  any    any     anywhere             anywhere             recent: UPDATE seconds: 7200 name: fail2ban-sshd side: source mask: 255.255.255.255 reject-with icmp-admin-prohibited
    0     0 hashlimit-syn-flood  tcp  --  any    any     anywhere             anywhere             tcp flags:FIN,SYN,RST,ACK/SYN
    0     0 LOG        all  --  any    any     anywhere             anywhere             recent: UPDATE seconds: 7200 name: badguys side: source mask: 255.255.255.255 limit: avg 2/min burst 5 LOG level warning ip-options prefix "ipt-recent-badguys: "
    0     0 REJECT     all  --  any    any     anywhere             anywhere             recent: UPDATE seconds: 7200 name: badguys side: source mask: 255.255.255.255 reject-with icmp-admin-prohibited
    0     0            all  --  any    any     anywhere             anywhere             recent: REMOVE name: badguys side: source mask: 255.255.255.255
    0     0 filter-ssh  tcp  --  any    any     anywhere             anywhere             tcp dpt:ssh
    0     0 ACCEPT     tcp  --  any    any     anywhere             anywhere             multiport dports http,https
    0     0 debops-legacy-input-rules  all  --  any    any     anywhere             anywhere
    0     0 REJECT     udp  --  any    any     anywhere             anywhere             reject-with icmp-port-unreachable
    0     0 REJECT     tcp  --  any    any     anywhere             anywhere             reject-with tcp-reset
    0     0 REJECT     all  --  any    any     anywhere             anywhere             reject-with icmp-proto-unreachable

Chain FORWARD (policy DROP 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 DROP       all  --  any    any     anywhere             anywhere             ctstate INVALID
    0     0 ACCEPT     all  --  any    any     anywhere             anywhere             ctstate RELATED,ESTABLISHED

Chain OUTPUT (policy ACCEPT 10 packets, 568 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 DROP       all  --  any    any     anywhere             anywhere             ctstate INVALID
   95 14010 ACCEPT     all  --  any    any     anywhere             anywhere             ctstate RELATED,ESTABLISHED

Chain ansible_controller-rules (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     tcp  --  any    any     250-233-149-118.nzcomms.co.nz  anywhere
    0     0 ACCEPT     tcp  --  any    any     52.183.xxx.xxx       anywhere

Chain debops-legacy-input-rules (1 references)
 pkts bytes target     prot opt in     out     source               destination

Chain filter-ssh (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 LOG        tcp  --  any    any     anywhere             anywhere             tcp dpt:ssh state NEW recent: UPDATE seconds: 300 hit_count: 8 name: ssh-new side: source mask: 255.255.255.255 limit: avg 2/min burst 5 LOG level warning ip-options prefix "ipt-recent-ssh-new: "
    0     0 REJECT     tcp  --  any    any     anywhere             anywhere             tcp dpt:ssh state NEW recent: UPDATE seconds: 300 hit_count: 8 name: ssh-new side: source mask: 255.255.255.255 reject-with icmp-admin-prohibited
    0     0            tcp  --  any    any     anywhere             anywhere             tcp dpt:ssh state NEW recent: SET name: ssh-new side: source mask: 255.255.255.255
    0     0 ACCEPT     tcp  --  any    any     anywhere             anywhere             tcp dpt:ssh

Chain hashlimit-icmp-flood (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     icmp --  any    any     anywhere             anywhere             limit: up to 10/sec burst 10 mode srcip htable-expire 3600000
    0     0 LOG        icmp --  any    any     anywhere             anywhere             limit: avg 2/min burst 5 LOG level warning ip-options prefix "ipt-hashlimit-icmp-flood: "
    0     0 DROP       icmp --  any    any     anywhere             anywhere

Chain hashlimit-syn-flood (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 RETURN     tcp  --  any    any     anywhere             anywhere             limit: up to 40/sec burst 40 mode srcip htable-expire 3600000
    0     0 LOG        tcp  --  any    any     anywhere             anywhere             limit: avg 2/min burst 5 LOG level warning ip-options prefix "ipt-hashlimit-syn-flood: "
    0     0 DROP       tcp  --  any    any     anywhere             anywhere
```

**Comment by drybjed** [comment]
So the ports are open... I'm trying to connect to that host, but can't at the moment. Is the host reachable over a public IP directly, or does it have a virtual router or gateway in front?

**Comment by jonathannz** [comment]
I am no master at reading and interpreting iptables, but I see no mention of 80 or 443 on there.

**Comment by jonathannz** [comment]
Should be reachable. It's been reachable for the deployment. I SSH into.

**Comment by inspiredearth** [comment]
BTW. `ferm` and `fail2ban` are back up. Had to, in order to get any info on the `iptables`

**Comment by inspiredearth** [comment]
Okay. So it's for sure working locally. I just did a request for the wordpress admin page, from on the server, and it gave what looks like the correct response.

```
--2017-07-14 09:52:15--  http://www.domain.nz/wp-admin
Resolving www.domain.nz (www.domain.nz)... 127.0.1.1
Connecting to www.domain.nz (www.domain.nz)|127.0.1.1|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: http://www.domain.nz/wp-admin/ [following]
--2017-07-14 09:52:15--  http://www.domain.nz/wp-admin/
Reusing existing connection to www.domain.nz:80.
HTTP request sent, awaiting response... 302 Found
Location: http://www.domain.nz/wp-login.php?redirect_to=http%3A%2F%2Fwww.domain.nz%2Fwp-admin%2F&reauth=1 [following]
--2017-07-14 09:52:15--  http://www.domain.nz/wp-login.php?redirect_to=http%3A%2F%2Fwww.domain.nz%2Fwp-admin%2F&reauth=1
Reusing existing connection to www.domain.nz:80.
HTTP request sent, awaiting response... 200 OK
Length: 2514 (2.5K) [text/html]
Saving to: ‘wp-admin’

wp-admin                                100%[===================
```

So it would seem that it's only from the public that it gives no response.

**Comment by inspiredearth** [comment]
When I do a `wget` from the debian workstation I did the deployment from, it gets no response. 
```
converted 'http://www.domain.nz/wp-admin' (ANSI_X3.4-1968) -> 'http://www.domain.nz/wp-admin' (UTF-8)
--2017-07-14 09:57:27--  http://www.domain.nz/wp-admin
Resolving www.domain.nz (www.domain.nz)... 52.183.xxx.xxx
Connecting to www.domain.nz (www.domain.nz)|52.183.xxx.xxx
```

So, as far as I can tell, the server stack is working fine (well, fine  enough to dish up the correct output when wp-login.php is requested, locally.

But it can not respond to requests from the public. As the firewall is not blocking them, I am going to take a punt and say it's something on the Azure configuration.

**Comment by inspiredearth** [comment]
Perhaps there's something more I need to do on the Azure configuration to make the public IP parse through to the server? Like setting up an endpoint for it? Looking into that now.

**Comment by drybjed** [comment]
Yes, it's usually called DMZ or port forwarding. Remember to forward HTTP (80) and HTTPS (443) ports.

**Comment by inspiredearth** [comment]
So I set up rules in the security group for the resource group that the VM is a part of. Allowing access to 80 and 443. But it seems to have made no difference. I am looking into why that might be the case.

**Comment by drybjed** [comment]
I can reach the website now:
```
POSTED ON JULY 12, 2017
Hello world!
Welcome to WordPress. This is your first post. Edit or delete it, then start writing!
```

**Comment by drybjed** [comment]
Here's a better debug output:
```
curl -sv [DATA EXPUNGED] >| /dev/null
* Hostname was NOT found in DNS cache
*   Trying [DATA EXPUNGED]...
* Connected to [DATA EXPUNGED] ([DATA EXPUNGED]) port 80 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.38.0
> Host: [DATA EXPUNGED]
> Accept: */*
> 
< HTTP/1.1 200 OK
* Server nginx is not blacklisted
< Server: nginx
< Date: [DATA EXPUNGED], 14 Jul 20[DATA EXPUNGED] 10:20:30 GMT
< Content-Type: text/html; charset=UTF-8
< Content-Length: [DATA EXPUNGED]
< Connection: keep-alive
< Vary: Accept-Encoding
< Vary: Accept-Encoding
< Age: 105
< X-Cache: HIT
< X-Cache-Hits: 4
< Accept-Ranges: bytes
< X-Clacks-Overhead: GNU Terry Pratchett
< 
{ [data not shown]
* Connection #0 to host [DATA EXPUNGED] left intact
```

**Comment by inspiredearth** [comment]
Oh boy. Deep long sigh... a big smile. 
Holy toledo Batman. There's actually a web site lurking in there!!

**Comment by drybjed** [comment]
Great. Now, destroy this machine and rebuild it again, just because you can.

**Comment by inspiredearth** [comment]
It's taken me nearly a week to build it this first time around. And the cool thing is, I'd be able to do it the second time (well, strictly speaking it would be the third time, as I already destroyed the first VM inadvertently)... in under 30 minutes. Basically, however long it takes for debops to do its thing.

Nice.

**Comment by jonathannz** [comment]
To ensure it doesn't create some kind of security vulnerability down the road, would you please remove the true domain references.

I can't tell you how great it is to break through to the other side of this. :-)

**Comment by inspiredearth** [comment]
And I am exceedingly grateful that you've hung in there with me these past 5 days. Thank you very much. I suspect I would have bailed days ago, and just made do with a Bitnami instance or something, had you not assisted as you have. Much appreciated.

**Comment by drybjed** [comment]
Sure, no problem. Cases like this help me find and fix bugs in DebOps, so it's certainly valuable. This time the reason was outside of the project's control though. I imagine that if you had set up DMZ earlier, it would be ready to go just after `debops wordpress` finished its run.

**Comment by inspiredearth** [comment]
100% correct. Once I got `debops wordpress` to run successfully, the only issue was on the Azure end of things. In all the stuff I read on setting up a VM on there, not once did I see any mention of setting up `endpoints` and `inbound rules` in the `network security group`.
I only stumbled across that when I found a large number of other people posting questions about why their server couldn't be access from the public IP. Then the penny dropped. :-)

**Comment by inspiredearth** [comment]
Although, mind you, I've learnt more than I knew about a few basic things to do with varnish, nginx, etc., whilst barking up the wrong tree trying to fix them.

**Comment by jonathannz** [comment]
Well, now that I am a master debops hacker (lol) I am going to sink my teeth into figuring out how to strap three more instances of Wordpress on the back of this beast.
I am assuming I just add their domains to the host file. Or will that simply map additional domains to the one Wordpress site?

**Comment by jonathannz** [comment]
Do you have any opinion on the following: (I ask, because I've read so many mixed and inconclusive views on it)

•  I have four wordpress sites. Each belongs to a branch of an organisation. One per country in which the organisation operates (it's a non-profit). 

• I will be the one managing the infrastructure for them all, from my master control centre in a base on the dark side of the moon. That includes managing much of the wp-admin stuff on each. Such as what plug-ins are installed, etc. The organisational branches themselves will simply manage their content.

• I was thinking it may be best to use Multisite. I have no experience with it, but plenty of experience with WP itself. My only concern is that I'll keep stumbling upon plug-ins that don't handle it. But I couldn't find any info on what proportion of plugins do and don't support it. Just that some won't.

I have no idea if WP is something you work with, so if this question is not something you can say much on, I understand.

**Comment by inspiredearth** [comment]
Okay, so I've decided to jump into WP MS after all.

Am I correct in understanding that if I make changes to the debops configuration such that it's set up for deploying a multisite wordpress installation, and then I deploy it... it will wipe out the existing WP installation, and replace it with a multisite installation? That's what I want it to do, and how I've understood it thus far.

**Comment by drybjed** [comment]
I haven't looked into the `debops-wordpress` custom roles themselves, you would need to ask @carlalexander about that. I imagine that you can just spin up a test VM and see for yourself what happens.

With DebOps you could move the WordPress database to a separate host, along with some filesystem space mounted over NFS, so that you can have mulitple WordPress worker nodes in front of the database, and behind some load balancer (`nginx`, `haproxy`). You could then deploy each site in a separate VM and route the connections using the load balancer. No need to fight with Multisite deployment that way.

**Comment by inspiredearth** [comment]
I tried converting the existing VM to multisite. Wasn't able to get it to work. The database had an issue of missing tables after I went through the process. So I've created a new VM and starting from scratch. Putting my new-found debops deployment powers to the test again. :-)

**Comment by inspiredearth** [comment]
Any thoughts on why I'd run into this error on the `debops` command? I ran the `debops bootstrap -u jonathan -b -K` first, and was asked for sudo password, which I provided. That run without errors. Then I did the `debops` and hit this:

```
TASK [debops.core : Install required core packages] *******************************************************************************************
failed: [www.primarydomain.nz] (item=libcap2-bin) => {"failed": true, "item": "libcap2-bin", "module_stderr": "Shared connection to www.DOMAIN.nz closed.\r\n", "module_stdout": "sudo: a password is required\r\n", "msg": "MODULE FAILURE", "rc": 1}
	to retry, use: --limit @/home/jonathan/.local/share/debops/debops-playbooks/playbooks/site.retry
```

I didn't have any password related issues when running these commands previously (on the 2nd VM).

**Comment by inspiredearth** [comment]
All good. I decided to add the user credentials onto the `debops` command, even though it wasn't necessary previously. Seems to be running okay now. `debops -u jonathan -b -K`.

**Comment by inspiredearth** [comment]
@drybjed Maciej, are you able to instruct me on how to set up debops for the domains nginx should watch for? Or point me to where I can find information on that?

For some reason the debops-wordpress deployment of Wordpress Multisite didn't push across any info to nginx configuration about the various networked site domains I specified. I create an [issue](https://github.com/carlalexander/debops-wordpress/issues/173) for this, but am not sure how long it'll take to get an answer.

I am thinking there must be a way I can just push the required nginx config changes through.

**Comment by drybjed** [comment]
@jonathannz The `wordpress` role from `debops-wordpress` defines the [`wordpress__nginx__server_name`](https://github.com/carlalexander/debops-wordpress/blob/master/roles/wordpress/defaults/main.yml#L360-L363) variable which contains the list of hostnames `nginx` will listen for. You should be able to redefine it in the Ansible inventory to override it with your own values.

**Comment by inspiredearth** [comment]
Thanks @drybjed . Reading your message, I realised my mistake. I didn't have the correct number of underscores in the variable. Had all sets of `__` where as the last one is single `_`. So I'd say that's why Nginx configuration was incomplete. I am rerunnign `debops wordpress` with the correct setting. I am assuming that will update it without wiping it out completely (as I've already access WP and set up the network sites). We'll see.

**Comment by inspiredearth** [comment]
Beautiful. Now they are all in the Nginx `.conf` file.

However, for some reasons it's render the server non-response to requests (both internally, and from public).  Eh...    `ERR_EMPTY_RESPONSE`


**Update**: RESOLVED for the main domain, **_but not for the other domains._**

**Comment by drybjed** [comment]
@jonathannz After googling a bit, it sounds like you need to configure the list of domain names in wordpress itself as well, so that the backend knows what domain to answer for. Perhaps look in that direction?

**Comment by inspiredearth** [comment]
Thanks. As far as I can tell, the Wordpress site is done. But I'll take another look over the various instructions I was following.

UPDATE: Well, I've gone over everything multiple times, and still none the wiser. DNS settings have propagated, correct domains are in Nginx conf, domains are set up in Wordpress MS, and wp-config.php is configured. Strange.

**Comment by drybjed** [comment]
@jonathannz So, it works now or not?

**Comment by inspiredearth** [comment]
Looks like we cross-posted. 
The main domain works. See message above. Networked sites are giving `ERR_EMPTY_RESPONSE`.

**Comment by inspiredearth** [comment]
Well, time to call it a day here. The server produces no response for the networked sites, both when request locally from the server, and from external (public). As far as I can tell, everything is configured correctly. I've tried, and re-tried everything I can think of. Time for a rest. Maybe I'll have  a bright idea in the morning. :-)

**Comment by carlalexander** [comment]
Sorry you had so many issues @jonathannz 😞 I haven't done any testing with DebOps and Azure so that's probably why there's so many kinks. Is the network site issue the same as #173?

**Comment by inspiredearth** [comment]
Hi Carl. Yes I set up a separate thread specific to the multisite issue, as this thread become a rather long handling of many issues @drybjed was kindly helping me with. I've answered your response to #173, and suggest we continue on there. Thanks very much.

**Comment by inspiredearth** [comment]
My hunch is that `err_empty_response` is likely a networking related error, as opposed to an issue in the application or the stack.  Although I could be wrong. Either way, I have no idea what else to try in order to fix it.

**Comment by inspiredearth** [comment]
Although the end result is not ideal (lack of SSL support), the server is now up and running, with the primary and networked domains and sites all functional. Closing this ticket.
